### ISSUE TITLE
Some questions.

### ISSUE THREAD

**Issue by brandon-reinhart** [bonsairobo/grid-tree-rs#4]
Some of these questions are driven from lack of understanding and only now using this library in a project. I'm working through that now to try and derive more understanding. For the sake of the questions, we'll say we're using a quadtree over a 2d space.

1. In general, why would I want to prefer signed vs unsigned indices?

2. The tree's root_nodes is a hashmap and so more than 4 entries could be placed in here, correct?

3. So then, for a grid of chunks, you expect to want to store many chunks in the root of the tree structure?

If those questions make sense, then is the idea here like brickmap? The root level is a grid, the children of that grid are trees, and the leaf nodes contents can be whatever. They could even be bricks, such as arrays of pixel material data pointing to a different type of subdivision strategy?

Then, if my understanding is correct, we might want to prefer signed indices in order to spatially insert newly loaded chunks to the spatial "left" (or negative direction) of existing ones, while preserving existing coordinates?

A root node can have a parent, which is itself an alloc_ptr, so is the idea that you could at runtime parent a root to a another tree if needed (for maybe dynamically loading a larger space and attaching the existing one to that)? Would you then want to, at some point, flatten this or unify it into a single grid tree? Or is there some other imagined use-case for the parent_ptr of a root node?

Given the parent is an alloc_ptr, does this mean I could parent a tree to a middle-layer of some other tree? Maybe to enable rapid edits of the world by just swapping those pointers instead of changing all of the leaves?

Thanks!

**Comment by bonsairobo** [comment]
@brandon-reinhart 

1. Depends on if you know your boundaries or not. If you know that your entire tree fits in one 3D bounding box, then you can safely use unsigned coordinates. Otherwise you will need signed coordinates.
2. Yes.
3. You have the option of storing chunk data in any node (or not if you use `Option<Chunk>`). The way I use it, the answer is yes.

> then is the idea here like brickmap?

Close but not really. In a brickmap, the voxels themselves can be pointers. `grid-tree` doesn't encourage this, since the pointers are stored separately from the node data, and the node pointers can be a different shape from the data chunks. A `grid-tree` pointer is a `u32`.

The reasoning is that I want to be able to store downsampled SDF data in the chunks. In order to have smooth transitions between levels of detail, it's nice to have a small branching factor (# children per node) like 2^3.

Technically other branching factors can be implemented with the `BranchShape` trait. Just keep in mind the branch nodes (not leaves) will take up `num_children * 4` bytes to store the pointers.

> Given the parent is an alloc_ptr, does this mean I could parent a tree to a middle-layer of some other tree?

Yes.

> Would you then want to, at some point, flatten this or unify it into a single grid tree? Or is there some other imagined use-case for the parent_ptr of a root node?

Yea the intent is to eventually merge the parented root into the other tree, starting at its parent.

> Maybe to enable rapid edits of the world by just swapping those pointers instead of changing all of the leaves?

Yea pretty much. The idea was to have and entire `Tree` dedicated to write-out-of-place edits that later gets merged into the "main" tree in a synchronized way. This would still require creating _new_ nodes in the main tree rather than just swapping pointers; _child_ pointers are only valid within a single `Tree`.

**Comment by brandon-reinhart** [comment]
Thanks, helpful to have my read of the code confirmed and you've clarified a lot!

For uploading the representation to the GPU, it seems like for each level I would want a pointer-to-children buffer and a values buffer. Level 0 would not need pointers to children.

I considered that the NodeAllocators members aren't pub, so that would prevent me from using the slabs to initialize the buffers. What's your recommended way of getting the raw representation out so it can be uploaded to the GPU? My intent is to try and use this representation in a compute shader.

Hmm. I may only want to upload from a certain root, so I guess uploading the entire allocator would be a bad idea. There might be "bucketload of stuff on disk -> bunch of stuff in RAM -> some stuff on GPU" where each level is a subset of the previous one.

**Comment by bonsairobo** [comment]
Ah yea I guess to upload to GPU you'd need access to the allocator buffers. Maybe you should fork this repo and do some experiments with that idea; I haven't tried it yet.

What's your plan for the compute shader? Are you trying to do ray-casting or something? You might want to keep the entire tree of pointers on GPU and then upload the data chunks as necessary for LoD. I'm not sure what the best strategy is for synchronizing the tree between CPU and GPU.

**Comment by brandon-reinhart** [comment]
Just thinking through how I might represent and render a Terraria style world, but with some more elaborate rendering. This seems like a pretty good way to handle the representation. 

For rendering I've had a few ideas:

1. I could mesh a chunk entirely using something like a greedy algorithm.
2. I could instance tiles by LOD size and draw those.
3. I could use a geometry shader to mesh on the fly.
4. I could push the tree to the gpu and walk it in a compute shader, per pixel.

1 & 2 would require me to stitch a texture together.
3 isn't an option for me atm, since wgpu doesn't support geometry shaders.
4 is interesting because I want to have the tree on the GPU anyway for tracing light & shadow and possibly for some collisions.

The other thing about 4 is that if the contents is a u32, it could be an index into sdf data. Or as an index + some flag bits. So, I'd think of the quadtree as a way to spatially partition SDFs that aren't moving relative to the world and which are fully contained within node bounds. Like tiles, or a segment of a water sim. Or maybe a tile might contain part of a cellular sim.

I'm also curious if by doing #4, I could have tiles that have light properties that would allow for caustics or at least something neat when light passes through the tile.

So that's all just guessing, but I'm working on building a few different methods to compare and measure them. I tend to take my time and write a lot of naive code as I learn the limits of particular ways of doing things.

I probably need something else for dynamic entities, like a bvh I can build quickly each frame. I'm still reading about that.

If I take the compute heavy route, my goal is that for any given pixel I can quickly determine which SDFs are relevant, which lights are relevant, etc. Like trying to figure out clustered shading but for 2d.

**Comment by brandon-reinhart** [comment]
I got this working on the GPU and its pretty fun. Rendering a quadtree tilemap with this representation in a compute shader is pretty much invariant to the size of the tree (the depth matters a little) and I'm mainly limited by what I can fit in memory. Which is easy enough to deal with when I add proper chunk streaming.

For now, I made the allocators pub and just throw them all into a couple buffers and then provide a buffer of layer offsets. This might not be ideal when it comes to editing, but I want to do some work on shading the tiles before I mess with editing. 

I haven't implemented any ray casting into the tree for lighting, but I don't see why that would be a problem.

Cool stuff!
